src/code/ucode_disas.c

```diff
@@ -0,0 +1,324 @@
+#include <global.h>
+
+#define F3DZEX_CONST(name) \
+    { name, #name }
+#define F3DZEX_FLAG(set, unset) \
+    { set, #set, #unset }
+#define F3DZEX_RENDERMODE(name, mask) \
+    { #name, name, mask }
+#define F3DZEX_SETRENDERMACRO(name, shift, len, value0, value1, value2, value3) \
+    {                                                                           \
+        name, shift, len, {                                                     \
+            { #value0, value0 }, { #value1, value1 }, { #value2, value2 }, {    \
+#value3, value3                                                 \
+            }                                                                   \
+        }                                                                       \
+    }
+
+#define DISAS_LOG        \
+    if (this->enableLog) \
+    osSyncPrintf
+
+u32 UCodeDisas_TranslateAddr(UCodeDisas* this, u32 addr) {
+    u32 physical = this->segments[SEGMENT_NUMBER(addr)] + SEGMENT_OFFSET(addr);
+    return PHYSICAL_TO_VIRTUAL(physical);
+}
+
+F3dzexConst sUCodeDisasGeometryModes[] = {
+    F3DZEX_CONST(G_ZBUFFER),     F3DZEX_CONST(G_TEXTURE_ENABLE),
+    F3DZEX_CONST(G_SHADE),       F3DZEX_CONST(G_SHADING_SMOOTH),
+    F3DZEX_CONST(G_CULL_FRONT),  F3DZEX_CONST(G_CULL_BACK),
+    F3DZEX_CONST(G_FOG),         F3DZEX_CONST(G_LIGHTING),
+    F3DZEX_CONST(G_TEXTURE_GEN), F3DZEX_CONST(G_TEXTURE_GEN_LINEAR),
+    F3DZEX_CONST(G_LOD),
+};
+
+F3dzexFlag sUCodeDisasMtxFlags[] = {
+    F3DZEX_FLAG(G_MTX_PROJECTION, G_MTX_MODELVIEW),
+    F3DZEX_FLAG(G_MTX_LOAD, G_MTX_MUL),
+    F3DZEX_FLAG(G_MTX_PUSH, G_MTX_NOPUSH),
+};
+
+const char* UCodeDisas_ParseCombineColor(u32 value, u32 idx) {
+    const char* ret = "?";
+
+    switch (value) {
+        case G_CCMUX_COMBINED:
+            ret = "COMBINED";
+            break;
+        case G_CCMUX_TEXEL0:
+            ret = "TEXEL0";
+            break;
+        case G_CCMUX_TEXEL1:
+            ret = "TEXEL1";
+            break;
+        case G_CCMUX_PRIMITIVE:
+            ret = "PRIMITIVE";
+            break;
+        case G_CCMUX_SHADE:
+            ret = "SHADE";
+            break;
+        case G_CCMUX_ENVIRONMENT:
+            ret = "ENVIRONMENT";
+            break;
+        case 6: {
+            ret = (idx == 2) ? "CENTER" : (idx == 3) ? "SCALE" : "1";
+            break;
+        }
```

Comment:
This applies to quite a few places in the file but imo braces should be avoided for cases, unless necessary for block variable declarations.
```suggestion
        case 6:
            ret = (idx == 2) ? "CENTER" : (idx == 3) ? "SCALE" : "1";
            break;
```